### PR TITLE
drpc: add deletion progression status

### PR DESCRIPTION
**Pull request by rakeshgm**
add drpc progression status as deleting
when drpc is being deleted

[BZ: 2136413](https://bugzilla.redhat.com/show_bug.cgi?id=2136413)

### CODE REVIEW

**Comment by rakeshgm** [comment]
I will it once and move from draft

**Comment by BenamarMk** [comment]
> I think this sends the wrong meaning. So adding a new phase like Delete and setting progression to Deleting when deleting DRPC makes sense.

@rakeshgm I think it is not necessary.  The way I look at it is that; it is Deployed/FailedOver/Relocated, but NOW we are deleting whatever we have deployed/FailedOver/Relocated.

File: controllers/drplacementcontrol_controller.go [hunk 1]
```diff
@@ -637,7 +637,13 @@ func (r *DRPlacementControlReconciler) Reconcile(ctx context.Context, req ctrl.R
 	if isBeingDeleted(drpc, placementObj) {
 		// DPRC depends on User PlacementRule/Placement. If DRPC or/and the User PlacementRule is deleted,
 		// then the DRPC should be deleted as well. The least we should do here is to clean up DPRC.
-		return r.processDeletion(ctx, drpc, placementObj, logger)
+		err := r.processDeletion(ctx, drpc, placementObj, logger)
+		if err != nil {
+			// update drpc progression only on err
+			logger.Info(fmt.Sprintf("Error in deleting DRPC: (%v)", err))
+			return r.setProgressionToDelete(drpc, logger)
```
Comment (BenamarMk) [inline controllers/drplacementcontrol_controller.go:644]: Maybe change the function to: `setProgressionAndUpdate`
That way, you can reuse the function `setProgression` and then just call update if the progression has changed. What I am trying to say is this:
```
setProgressionAndUpdate(ProgressionSetDeleting)
...
...
func (r *DRPlacementControlReconciler) setProgressionAndUpdate(...) {
    setProgression(...)
    Update(...)
}
```

File: controllers/drplacementcontrol_controller.go [hunk 2]
```diff
@@ -659,6 +665,25 @@ func (r *DRPlacementControlReconciler) Reconcile(ctx context.Context, req ctrl.R
...
+
+	progressionDeleting := rmn.ProgressionSetDeleting
+	if drpc.Status.Progression != progressionDeleting {
+		logger.Info(fmt.Sprintf("Progression: Current '%s'. Next '%s'",
+			drpc.Status.Progression, progressionDeleting))
+
+		drpc.Status.Progression = progressionDeleting
+	}
+
+	if err := r.Status().Update(context.TODO(), drpc); err != nil {
```
Comment (BenamarMk) [inline controllers/drplacementcontrol_controller.go:681]: You don't need to call `Update` if no change to the progression.

File: controllers/drplacementcontrol_controller.go [hunk 3]
```diff
@@ -635,6 +645,20 @@ func (r *DRPlacementControlReconciler) Reconcile(ctx context.Context, req ctrl.R
 	return r.reconcileDRPCInstance(d, logger)
 }
 
+func (r *DRPlacementControlReconciler) setProgressionAndUpdate(
+	ctx context.Context, drpc *rmn.DRPlacementControl,
+) (ctrl.Result, error) {
+	updated := updateDRPCProgression(drpc, rmn.ProgressionSetDeleting, r.Log)
+
+	if updated {
+		if err := r.Update(ctx, drpc); err != nil {
```
Comment (BenamarMk) [inline controllers/drplacementcontrol_controller.go:654]: You want to make sure that the status is updated right?

Comment (rakeshgm) [reply controllers/drplacementcontrol_controller.go]: ok. corrected it now.

**Comment by rakeshgm** [comment]
The unit tests are passing. I am moving the PR from draft.

File: api/v1alpha1/drplacementcontrol_types.go [hunk 4]
```diff
@@ -92,6 +92,7 @@ const (
 	ProgressionUpdatedPlacement                    = ProgressionStatus("UpdatedPlacement")
 	ProgressionEnsuringVolSyncSetup                = ProgressionStatus("EnsuringVolSyncSetup")
 	ProgressionSettingupVolsyncDest                = ProgressionStatus("SettingUpVolSyncDest")
+	ProgressionSetDeleting                         = ProgressionStatus("Deleting")
```
Comment (BenamarMk) [inline api/v1alpha1/drplacementcontrol_types.go:95]: Why is it `ProgressionSetDeleting`? Why not `ProgressionDeleting` like the others? What does `Set` mean here?

File: controllers/drplacementcontrol_controller.go [hunk 5]
```diff
@@ -613,7 +615,15 @@ func (r *DRPlacementControlReconciler) Reconcile(ctx context.Context, req ctrl.R
 	if isBeingDeleted(drpc, placementObj) {
 		// DPRC depends on User PlacementRule/Placement. If DRPC or/and the User PlacementRule is deleted,
 		// then the DRPC should be deleted as well. The least we should do here is to clean up DPRC.
-		return r.processDeletion(ctx, drpc, placementObj, logger)
+		err := r.processDeletion(ctx, drpc, placementObj, logger)
+		if err != nil {
+			// update drpc progression only on err
+			logger.Info(fmt.Sprintf("Error in deleting DRPC: (%v)", err))
+
+			return r.setProgressionAndUpdate(ctx, drpc, rmn.ProgressionDeleting)
```
Comment (BenamarMk) [inline controllers/drplacementcontrol_controller.go:623]: This function has nothing to do with `ctrl.Result{}`. I would make it return `nil` or `error`. That's all.  So in this line, I would change it to this:
```
return ctrl.Result{}, r.setProgressionAndUpdate(ctx, drpc, rmn.ProgressionDeleting)
```

**Review by BenamarMk — APPROVED** [review]

**Pull request merged**